core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSession.java

```diff
@@ -19,14 +18,14 @@ record EntityVariablePair<Solution_>(VariableDescriptor<Solution_> variableDescr
         @Override
         public boolean equals(@Nullable Object o) {
             if (o instanceof EntityVariablePair<?> other) {
-                return entity == other.entity && Objects.equals(variableDescriptor, other.variableDescriptor);
+                return entity == other.entity && variableDescriptor.getOrdinal() == other.variableDescriptor.getOrdinal();
             }
             return false;
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(System.identityHashCode(entity), variableDescriptor);
+            return (31 * System.identityHashCode(entity)) ^ variableDescriptor.getOrdinal();
```

Comment:
... but ordinal has the function. In either case, arguably we should put this logic inside `VariableDescriptor` and not externalize it here.

Reply:
And maybe I would add a comment here that this class is often used as key in hashmaps, and therefore this method needs to be optimized.